File: Data Structures & Algorithms/longest-increasing-subsequence/submission-2.java
class Solution {
    public int lengthOfLIS(int[] nums) {
        return lengthOfLISHelper(nums,nums.length,-1);
    }
    public int lengthOfLISHelper(int[] nums, int i, int prev_index){
        if(i==0) return 0;
        int not_taken = lengthOfLISHelper(nums,i-1,prev_index);
        int taken = 0;
        if(prev_index == -1 || nums[i-1] < nums[prev_index]){
            taken = 1 + lengthOfLISHelper(nums,i-1,i-1);
        }
        return Math.max(taken, not_taken);
    }
}

File: Data Structures & Algorithms/jump-game-ii/submission-2.py
class Solution:
    def jump(self, nums: List[int]) -> int:
        jumps = 0
        l = 0
        r = 0
        n = len(nums)
        while r < n-1:
            farthest = 0
            for i in range(l,r+1):
                farthest = max(i+nums[i],farthest)
            l = r+1
            r = farthest
            jumps += 1
        return jumps
        